### PR TITLE
[SPARK-13029][ml] fix a logistic regression issue when input data has a column with identical value

### PR DESCRIPTION
**Pull request by coderxiang**
This is a bug that appears while fitting a Logistic Regression model with and `setFitIntercept(false)`. If the data matrix has one column with identical value, the resulting model is often not correct. Specifically, the special column will always get a weight of 0, due to the logic that checks columns with std=0. However, the correct solution, which is unique for L2 logistic regression, usually has non-zero weight.

The fix is to update the special handing logic to make it compatible with columns that has 0 variance.

Two unit tests are included in this PR where one of them is for review only. I'll remove it if this is going to be merged.

cc @mengxr @dbtsai

### CODE REVIEW

**Comment by coderxiang** [comment]
Jenkins, retest please

**Comment by coderxiang** [comment]
Jenkins, retest please.

**Comment by coderxiang** [comment]
Jenkins, test this please.

File: mllib/src/test/scala/org/apache/spark/ml/classification/LogisticRegressionSuite.scala [hunk 1]
```diff
@@ -607,6 +607,73 @@ class LogisticRegressionSuite
     assert(model2.coefficients ~= coefficientsR2 relTol 1E-2)
   }
 
+
+  test("an extra large example for review only") {
```
Comment (coderxiang) [inline mllib/src/test/scala/org/apache/spark/ml/classification/LogisticRegressionSuite.scala]: This test is for review only, to show an example on a larger data set. Will remove if merged.

File: data/mllib/sample_libsvm_data_with_unique_column.txt [hunk 2]
```diff
@@ -0,0 +1,270 @@
+1 1:0.708333 2:1 3:1 4:-0.320755 5:-0.105023 6:-1 7:1 8:-0.419847 9:-1 10:-0.225806 12:1 13:-1 14:1
```
Comment (coderxiang) [inline data/mllib/sample_libsvm_data_with_unique_column.txt]: This file is for review only, to show an example on a larger data set. Will remove if merged.

**Comment by coderxiang** [comment]
Jenkins, test this please.

File: mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala [hunk 3]
```diff
@@ -971,8 +971,12 @@ private class LogisticAggregator(
           val margin = - {
             var sum = 0.0
             features.foreachActive { (index, value) =>
-              if (featuresStd(index) != 0.0 && value != 0.0) {
-                sum += localCoefficientsArray(index) * (value / featuresStd(index))
+              if (value != 0.0) {
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala]: This `if` branch is not necessary.

File: mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala [hunk 4]
```diff
@@ -983,8 +987,12 @@ private class LogisticAggregator(
           val multiplier = weight * (1.0 / (1.0 + math.exp(margin)) - label)
 
           features.foreachActive { (index, value) =>
-            if (featuresStd(index) != 0.0 && value != 0.0) {
-              localGradientSumArray(index) += multiplier * (value / featuresStd(index))
+            if (value != 0.0) {
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala]: same here

**Comment by mengxr** [comment]
@coderxiang Would the fix be cleaner if we set `featuresStd(i)` to `1.0` if it is `0.0`?

File: mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala (hunk 3 above)
Comment (coderxiang) [inline mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala]: Sure I'll remove, I was avoiding changing existing logic.

**Comment by coderxiang** [comment]
@mengxr you mean do this locally? I was concerned this will create confusion since we are modifying the true value.

**Comment by coderxiang** [comment]
Jenkins, test this please.

File: mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala [hunk 5]
```diff
@@ -971,8 +971,10 @@ private class LogisticAggregator(
           val margin = - {
             var sum = 0.0
             features.foreachActive { (index, value) =>
-              if (featuresStd(index) != 0.0 && value != 0.0) {
+              if (featuresStd(index) != 0.0) {
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala]: The previous version is correct. Checking `value != 0.0` is much cheaper than computing `localCoefficientsArray(index)`.

File: mllib/src/test/scala/org/apache/spark/ml/classification/LogisticRegressionSuite.scala [hunk 6]
```diff
@@ -607,6 +607,74 @@ class LogisticRegressionSuite
...
+    val model = trainer.fit(binaryDatasetWithUniqueColumn)
+
+    val interceptR = 0.0
+    val coefficientsR = Vectors.dense(0.22478867, -0.02241016)
+
+    assert(model.intercept ~== interceptR absTol 1E-3)
+    assert(model.coefficients ~= coefficientsR relTol 1E-2)
+
+    /*
+        Use the following scikit-learn Python code to get a reference result:
```
Comment (mengxr) [inline mllib/src/test/scala/org/apache/spark/ml/classification/LogisticRegressionSuite.scala]: Please move the comments to the beginning of the test.

File: mllib/src/test/scala/org/apache/spark/ml/classification/LogisticRegressionSuite.scala [hunk 7]
```diff
@@ -607,6 +607,74 @@ class LogisticRegressionSuite
...
+      0.0637250665085166, -0.0589532651377124, 0.1383368129434264,
+      0.0665749825701113, 0.0799386779781182, 0.1198682685242071,
+      0.1802933312643371, -0.0124797701753129)
+
+    assert(model1.intercept ~== interceptR1 absTol 1E-3)
+    assert(model1.coefficients ~= coefficientsR1 relTol 1E-2)
+  }
+
+  test("binary logistic regression without intercept with L2 regularizationon " +
+    "data with unique column without intercept") {
```
Comment (mengxr) [inline mllib/src/test/scala/org/apache/spark/ml/classification/LogisticRegressionSuite.scala]: `unique column` -> `a constant column`

File: mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala [hunk 8]
```diff
@@ -341,11 +341,11 @@ class LogisticRegression @Since("1.2.0") (
...
                 // If `standardization` is false, we still standardize the data
-                // to improve the rate of convergence; as a result, we have to
-                // perform this reverse standardization by penalizing each component
-                // differently to get effectively the same objective function when
+                // to improve the rate of convergence unless the standard deviation is zero;
+                // as a result, we have to perform this reverse standardization by penalizing
+                // each component differently to get effectively the same objective function when
                 // the training dataset is not standardized.
-                if (featuresStd(index) != 0.0) regParamL1 / featuresStd(index) else 0.0
+                if (featuresStd(index) != 0.0) regParamL1 / featuresStd(index) else regParamL1
```
Comment (dbtsai) [inline mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala]: The constant `value` can be really large or very small negatively. The optimizer may not be able to converge well in this case. I don't prove or try it yet, but mathematically, with the following changes, this should be solving identical problem.

```scala
// the training dataset is not standardized.
if (featuresStd(index) != 0.0) regParamL1 / featuresStd(index) else regParamL1 / featuresMean(index)
```

File: mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala [hunk 9]
```diff
@@ -417,7 +417,7 @@ class LogisticRegression @Since("1.2.0") (
         val rawCoefficients = state.x.toArray.clone()
         var i = 0
         while (i < numFeatures) {
-          rawCoefficients(i) *= { if (featuresStd(i) != 0.0) 1.0 / featuresStd(i) else 0.0 }
+          rawCoefficients(i) *= { if (featuresStd(i) != 0.0) 1.0 / featuresStd(i) else 1.0 }
```
Comment (dbtsai) [inline mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala]:
```scala
rawCoefficients(i) *= { if (featuresStd(i) != 0.0) 1.0 / featuresStd(i) else 1.0 / featuresMean(i) }
```

File: mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala [hunk 10]
```diff
@@ -971,8 +971,10 @@ private class LogisticAggregator(
           val margin = - {
             var sum = 0.0
             features.foreachActive { (index, value) =>
-              if (featuresStd(index) != 0.0 && value != 0.0) {
+              if (featuresStd(index) != 0.0) {
                 sum += localCoefficientsArray(index) * (value / featuresStd(index))
+              } else {
+                sum += localCoefficientsArray(index) * value
```
Comment (dbtsai) [inline mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala]: Change `value` into `1.0`

File: mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala [hunk 11]
```diff
@@ -983,8 +985,10 @@ private class LogisticAggregator(
           val multiplier = weight * (1.0 / (1.0 + math.exp(margin)) - label)
 
           features.foreachActive { (index, value) =>
-            if (featuresStd(index) != 0.0 && value != 0.0) {
+            if (featuresStd(index) != 0.0) {
               localGradientSumArray(index) += multiplier * (value / featuresStd(index))
+            } else {
+              localGradientSumArray(index) += multiplier * value
```
Comment (dbtsai) [inline mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala]: Change `value` into `1.0`

File: mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala [hunk 12]
```diff
@@ -1106,7 +1110,8 @@ private class LogisticCostFun(
                 totalGradientArray(index) += regParamL2 * temp
                 value * temp
               } else {
-                0.0
+                totalGradientArray(index) += regParamL2 * value
+                value * value
```
Comment (dbtsai) [inline mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala]: Change `value` into `1.0`

File: mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala (hunk 8 above)
Comment (coderxiang) [inline mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala]: Can you give an example that using `value` will fail to converge? 

I agree any non-zero number here can make the algorithm work, but should we select a particular number as the denominator, or let it be the original value? @mengxr what do you think?

Comment (dbtsai) [inline mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala]: BTW, this technique is already used here by dividing the std of feature. Note that you need to make sure `value != 0`, and if `value == 0`, the coefficient of that feature will be zero.

**Comment by coderxiang** [comment]
Jenkins, test this please.

**Comment by dbtsai** [comment]
Intuitively, this makes sense to me. Since when `setFitIntercept(false)`, those features with `std == 0` will act as the effect of intercept resulting non-zero coefficients. 

Can you add couple more tests as the following. Thanks.

First, adding two new datasets by zeroing out `binaryDataset` and making one column as non-zero constance. 

Matching the result against GLMNET like the rest of the tests when
1)  setFitIntercept(false), setStandardization(false) with/without regularization 
2)  setFitIntercept(false), setStandardization(true) with/without regularization
3)  setFitIntercept(true), setStandardization(false) with/without regularization
4)  setFitIntercept(true), setStandardization(true) with/without regularization

+cc @iyounus Linear Regression may have similar issue, if you have time, you may check it out. Thanks.

File: mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala (hunk 8 above)
Comment (dbtsai) [inline mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala]: @coderxiang You may try it out. When `value` is very small like `1e-6` as a constant comparing the rest of the features, the corresponding coefficient will be very large. In this case, the optimization on coefficients will be on different scales, and this often causes some convergence issue in line search. 

Similar argument can be made when the constant `value` is very large comparing the rest of the features. 

That's why we still standardize the features even users ask `standardization == false`.

**Comment by coderxiang** [comment]
@dbtsai without regularization may let the objective being not strongly-convex and thus not guaranteeing the uniqueness of the solution.

**Comment by coderxiang** [comment]
Jenkins, test this please.

**Comment by dbtsai** [comment]
@coderxiang I agree, without regularization, those features become collinear so the solution will not be unique. However, for those features with std != 0, the coefficients should be unique. Can you check them at least?

**Comment by iyounus** [comment]
@dbtsai Linear regression also has similar issues. There, "normal" and "l-gbfs" solvers treat this case differently (and incorrectly). The other problem there is that if intercept=true, then a constant feature column makes the gramian matrix singular and cholesky decomposition fails. Should I create separate jira for the case of constant feature?

**Comment by dbtsai** [comment]
@iyounus Ideally, it will be great that `intercept=true`, we keep the current behavior which is constant column doesn't have any predictive power.

**Comment by iyounus** [comment]
@dbtsai I agree that the constant feature doesn't have predictive power. But, the WeightedLeastSqures just throws an `AssertionError` in `lapack.dpotrs` (https://issues.apache.org/jira/browse/SPARK-11918), whereas the "l-bfgs" solver sets the coefficient to zero.

**Comment by dbtsai** [comment]
@iyounus Maybe in this case, `WeightedLeastSqures ` should drop those columns so the model can be still trained.

**Comment by mengxr** [comment]
@coderxiang @dbtsai Sorry for late response! I actually thought this PR already got merged ... Anyway, I tested `glmnet` and found that `glmnet` outputs zero coefficients for constant columns regardless of intercept, regularization, and standardization settings. I thought about it today and I feel it actually makes sense. If we have a constant column in our training data, do we expect it to change or stay constant in test data? If its value might change, we should set its coefficient to zero because we cannot estimate how big the change would be. If its value stays constant (or maybe users created this column to add bias manually), it shouldn't be regularized and users should really turn on `fitIntercept` instead. So my suggestion is to follow glmnet and set the coefficients of constant columns to zero regardless of other settings. If there are constant columns and `fitIntercept` is false. We should output a warning message. Does it sound good to you?

**Comment by mengxr** [comment]
Had an offline discussion with @dbtsai and @coderxiang . We agreed to keep the current behavior and have it well documented. I will mark this JIRA as "won't" and created SPARK-13590 for documentation and logging improvement.

@coderxiang Do you mind closing this PR?

**Pull request closed without merging**